### PR TITLE
GH-30717: [C#] Add ToString() methods to Arrow classes

### LINKED ISSUE #30717: [C#] Add ToString() methods to Arrow classes
We should add `ToString` methods to `RecordBatch`, `Schema`, `Field`, `DataType`, `Table`, and `ChunkedArray`.

The default implementation in C# is just to return the class name, which isn't very useful.

**Reporter**: [Will Jones](https://issues.apache.org/jira/browse/ARROW-15217) / @wjones127

<sub>**Note**: *This issue was originally created as [ARROW-15217](https://issues.apache.org/jira/browse/ARROW-15217). Please see the [migration documentation](https://github.com/apache/arrow/issues/14542) for further details.*</sub>

### PR DESCRIPTION
**Pull request by voidstar69**
### What changes are included in this PR?

Implemented the ToString() method on classes ChunkedArray, Field, RecordBatch, Schema and Table. 
I could not find the class DataType mentioned in the issue, perhaps it meant DateType?

Closes #30717.
* Closes: #30717

### CODE REVIEW

**Review by westonpace — CHANGES_REQUESTED** [review]
These are helpful but I don't know if they address the heart of the issue.  My guess is the issue was asking for some kind of pretty printing support.  Do you want to perhaps add a follow-up issue for actual table-printing support?  For example, here is what the "to string" looks like in various other implementations:

### Python:

```
pyarrow.Table
strings: string
ints: int32
floats: float
----
strings: [["one","two","three","four","five"]]
ints: [[1,2,3,4,5]]
floats: [[1,2,3,4,5]]
```

### C++

```
strings: string
ints: int32
floats: float
----
strings:
  [
    [
      "one",
      "two",
      "three",
      "four",
      "five"
    ]
  ]
ints:
  [
    [
      1,
      2,
      3,
      4,
      5
    ]
  ]
floats:
  [
    [
      1,
      2,
      3,
      4,
      5
    ]
  ]
```

### Rust:

```
RecordBatch { schema: Schema { fields: [Field { name: "strings", data_type: Utf8, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "ints", data_type: Int32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "floats", data_type: Float64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }, columns: [StringArray
[
  "one",
  "two",
  "three",
  "four",
  "five",
], PrimitiveArray<Int32>
[
  1,
  2,
  3,
  4,
  5,
], PrimitiveArray<Float64>
[
  1.0,
  2.0,
  3.0,
  4.0,
  5.0,
]], row_count: 5 }
```

**Comment by voidstar69** [comment]
> These are helpful but I don't know if they address the heart of the issue. My guess is the issue was asking for some kind of pretty printing support. Do you want to perhaps add a follow-up issue for actual table-printing support? For example, here is what the "to string" looks like in various other implementations:
> 
> ### Python:
> ```
> pyarrow.Table
> strings: string
> ints: int32
> floats: float
> ----
> strings: [["one","two","three","four","five"]]
> ints: [[1,2,3,4,5]]
> floats: [[1,2,3,4,5]]
> ```
> 
> ### C++
> ```
> strings: string
> ints: int32
> floats: float
> ----
> strings:
>   [
>     [
>       "one",
>       "two",
>       "three",
>       "four",
>       "five"
>     ]
>   ]
> ints:
>   [
>     [
>       1,
>       2,
>       3,
>       4,
>       5
>     ]
>   ]
> floats:
>   [
>     [
>       1,
>       2,
>       3,
>       4,
>       5
>     ]
>   ]
> ```
> 
> ### Rust:
> ```
> RecordBatch { schema: Schema { fields: [Field { name: "strings", data_type: Utf8, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "ints", data_type: Int32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "floats", data_type: Float64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }, columns: [StringArray
> [
>   "one",
>   "two",
>   "three",
>   "four",
>   "five",
> ], PrimitiveArray<Int32>
> [
>   1,
>   2,
>   3,
>   4,
>   5,
> ], PrimitiveArray<Float64>
> [
>   1.0,
>   2.0,
>   3.0,
>   4.0,
>   5.0,
> ]], row_count: 5 }
> ```

I see. I thought the original issue might have been about making these types easier to comprehend in a debugger, such as in Visual Studio. There the debugger invokes `ToString` on an object, which usually returns a single line of text which the debugger then displays. I think this current PR achieves that goal, although I cannot know the intent of the original issue.

I have now created issue #36777 for adding pretty printing methods.

File: csharp/src/Apache.Arrow/RecordBatch.cs [hunk 1]
```diff
@@ -93,5 +93,7 @@ public RecordBatch Clone(MemoryAllocator allocator = default)
             IEnumerable<IArrowArray> arrays = _arrays.Select(array => ArrowArrayFactory.BuildArray(array.Data.Clone(allocator)));
             return new RecordBatch(Schema, arrays, Length);
         }
+
+        public override string ToString() => $"{nameof(RecordBatch)}: Length={Length}, ColumnCount={ColumnCount}";
```
[the remark this answers is not in the document]
Comment (voidstar69) [reply csharp/src/Apache.Arrow/RecordBatch.cs]: Done, updated.

File: csharp/src/Apache.Arrow/Schema.cs [hunk 2]
```diff
@@ -114,5 +114,7 @@ public Schema SetField(int fieldIndex, Field newField)
 
             return new Schema(fields, Metadata);
         }
+
+        public override string ToString() => $"{nameof(Schema)}: Num fields={_fieldsList.Count}, Num metadata={Metadata?.Count}";
```
Comment (westonpace) [inline csharp/src/Apache.Arrow/Schema.cs:118]:
```suggestion
        public override string ToString() => $"{nameof(Schema)}: Num fields={_fieldsList.Count}, Num metadata={Metadata?.Count ?? 0}";
```

**Review by westonpace — CHANGES_REQUESTED** [review]
I agree it makes sense to put pretty printing somewhere else other than `ToString`.  One more tiny change if you don't mind.  Also, could you briefly add some calls to `ToString` in the unit tests for these types?  You probably don't need to create a new test but just something to make sure it converts to string without triggering an exception.

**Comment by voidstar69** [comment]
> I agree it makes sense to put pretty printing somewhere else other than `ToString`. One more tiny change if you don't mind. Also, could you briefly add some calls to `ToString` in the unit tests for these types? You probably don't need to create a new test but just something to make sure it converts to string without triggering an exception.

@westonpace, I think this PR is now ready for your review. Hopefully we can merge this PR soon. Thanks.

**Comment by voidstar69** [comment]
@westonpace are you able to review this PR now? Or is there somebody else who can do this? Thanks.

**Comment by voidstar69** [comment]
@westonpace

**Comment by voidstar69** [comment]
@davidhcoe are you or another maintainer available to review this PR please? This PR has not received a further review since @westonpace reviewed it in July. I have held off from creating further PRs due to this.

**Comment by CurtHagenlocher** [comment]
Sorry, meant to look last week but I couldn't figure out how to re-run the failed appveyor build. As I can't imagine it's related, I'll ignore it and cross my fingers... .

**Pull request merged**